osquery/events/windows/ntfs_event_publisher.cpp

```diff
@@ -0,0 +1,595 @@
+  // drive letter in order to restart the reader services
+  Config::get().files(
+      [&configuration](const std::string& category,
+                       const std::vector<std::string>& path_list) {
+        for (const auto& path : path_list) {
+          const auto& drive_letter = static_cast<char>(::toupper(path.front()));
+          configuration.insert(drive_letter);
+        }
+      });
+
+  return configuration;
+}
+
+Status NTFSEventPublisher::getPathFromReferenceNumber(
+    std::string& path, char drive_letter, const USNFileReferenceNumber& ref) {
+  path.clear();
+
+  // Get the root folder handle
+  VolumeData volume_data = {};
+  auto status = getVolumeData(volume_data, drive_letter);
+  if (!status.ok()) {
+    return status;
+  }
+
+  // See if we have been requested to solve the root folder
+  if (ref == volume_data.root_ref) {
+    path.push_back(drive_letter);
+    path.append(":\\");
+
+    return Status(0);
+  }
+
+  // Convert the reference number to the native Windows structure
+  FILE_ID_DESCRIPTOR native_file_id = {};
+  GetNativeFileIdFromUSNReference(native_file_id, ref);
+
+  // Attempt to open the file or folder
+  auto handle =
+      OpenFileById(volume_data.volume_handle,
+                   &native_file_id,
+                   0,
+                   FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                   nullptr,
+                   FILE_FLAG_BACKUP_SEMANTICS);
+  if (handle == INVALID_HANDLE_VALUE) {
+    std::stringstream message;
+    message << "Failed to open the file in volume " << drive_letter
+            << ":\\. Error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, ::GetLastError())) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status(1, message.str());
+  }
+
+  auto required_bytes = static_cast<size_t>(::GetFinalPathNameByHandle(
+      handle, nullptr, 0, FILE_NAME_NORMALIZED | VOLUME_NAME_DOS));
+
+  if (required_bytes == 0U) {
+    auto error_code = ::GetLastError();
+    ::CloseHandle(handle);
+
+    std::stringstream message;
+    message << "Failed to determine the path size for the file in volume "
+            << drive_letter << ":\\. Error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, error_code)) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status(1, message.str());
+  }
+
+  // We are going to add an additional byte, as we may or may not have the null
+  // terminator already included depending on the operating system version
+  std::string buffer;
+  required_bytes += 1U;
+
+  buffer.resize(required_bytes);
+  if (buffer.size() != required_bytes) {
+    ::CloseHandle(handle);
+    throw std::bad_alloc();
+  }
+
+  auto bytes_returned = static_cast<size_t>(
+      ::GetFinalPathNameByHandle(handle,
+                                 &buffer[0],
```

Comment:
does `buffer.data()` work?

Reply:
Ah, looks like it doesn't: `GetFinalPathNameByHandle` needs an `LPSTR` (i.e. `char *`), but `data()` returns `const char *`.